### DISCUSSION TITLE
Debugger

### DISCUSSION THREAD

**Discussion by KevDev001** [ScratchAddons/ScratchAddons#4118] [Help] [answered]
How does the scratch debugger work?

**Comment by CST1229** [comment]
You mean our debugger addon? Which part of it? (It's very complicated)

**Reply by KevDev001** [reply]
Yeah, I'm just a bit confused on how to use it and how it works.

**Comment by Chiroyce1** [comment, marked as answer]
### Introduction to Debuggers
Debuggers are used for most programming languages, it helps programmers to find and fix bugs more easily. With the help of a debugger, you can step through individual lines of code, see the values of variables, etc, to try to find where an error/bug occurs.

Scratch was meant for beginners - so adding a debugger wasn't considered by the Scratch Team, so the wonderful people at Scratch Addons decided to make one.

To enable it, head over to the settings, search for "debugger" and make sure the switch is orange (click it if it isn't).

### Demo of the Scratch Addons Debugger
To test it out, head over to fresh new Scratch project, and add some demo code like this —
<img width="553" alt="image" src="https://user-images.githubusercontent.com/97374054/152825887-cf36f462-00e9-45a9-bd99-6db2aeda8f68.png">

Now, if you click the green flag, everything happens so fast, that you can barely see the variable increase, here the debugger will help you!

So to begin debugging, our task will be to find out what the random number is, without using any extra variable or say blocks. This can be useful if you want to see a lot of values in detail without spending too much time on it. 

From the "Debugger" category in the block palette, drag the "breakpoint" block and place it under the `change my variable` block like this

<img width="521" alt="image" src="https://user-images.githubusercontent.com/97374054/152826574-bef77501-19af-4982-8b1d-27876791fade.png">

Now if you click the green flag, you should see the debugging window appear, now look at the top section of it, right now we're only gonna focus on the `resume` button. Before that, take a look at the value of `my variable` on the stage. This was our first random number, we were able to find it because the program stopped for us to see it. If you press `resume`, the variable will change again. Our random number this time will be `(current_number) - (first_number)`, if you click it again, you can see the number rise up randomly. This is how the breakpoint system works. I'll try to explain the other tabs soon.

**Reply by WorldLanguages** [reply]
A video explaining how to use the debugger would be great for our YouTube channel
